internal/aws/awsutil/conn.go

```diff
@@ -203,29 +203,58 @@ func GetAWSConfigSession(logger *zap.Logger, cn ConnAttr, cfg *AWSSessionSetting
-		config.Credentials = credentials.NewCredentials(&credentials.ChainProvider{
-			Providers: customCredentialProvider(cfg, config),
-		})
-	}
 	return config, s, nil
 }
 
+// order
+// 1. shared profile
+// 2. env var
+// 3. shared creds file
+// 4. ec2 creds
+// 5. override creds providers
+// explicit keys do not make sense in this context since contrib does not take in explicit keys
+// and cwa does not provide explicit keys to config but come from these methods of getting creds
+// @TODO make this more upstream friendly working in the future
 func customCredentialProvider(cfg *AWSSessionSettings, config *aws.Config) []credentials.Provider {
-	defaultCredProviders := defaults.CredProviders(config, defaults.Handlers())
 	overrideCredProviders := override.GetCredentialsChainOverride().GetCredentialsChain()
 	credProviders := make([]credentials.Provider, 0)
-	// if is for differently configured shared creds file location
-	// else if is for diff profile but no change in creds file ex run in containers
-	if cfg.SharedCredentialsFile != nil && len(cfg.SharedCredentialsFile) > 0 {
+	// DO NOT TRY TO OPTIMIZE THIS CODE !!!
+	// If you do please do not change order
+	// this is to keep ordering where if there is a profile we want to try those creds first
+	if cfg.Profile != "" && cfg.SharedCredentialsFile != nil && len(cfg.SharedCredentialsFile) > 0 {
 		for _, file := range cfg.SharedCredentialsFile {
-			credProviders = append(credProviders, &credentials.SharedCredentialsProvider{Filename: file, Profile: cfg.Profile})
+			credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{
+				sharedCredentialsProvider: &credentials.SharedCredentialsProvider{
+					Filename: file,
+					Profile:  cfg.Profile,
+				},
+			})
 		}
 	} else if cfg.Profile != "" {
-		credProviders = append(credProviders, &credentials.SharedCredentialsProvider{Filename: "", Profile: cfg.Profile})
+		credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{
+			sharedCredentialsProvider: &credentials.SharedCredentialsProvider{
+				Filename: "",
+				Profile:  cfg.Profile,
+			},
+		})
+	}
+	credProviders = append(credProviders, &credentials.EnvProvider{})
+	if cfg.Profile == "" && cfg.SharedCredentialsFile != nil && len(cfg.SharedCredentialsFile) > 0 {
+		for _, file := range cfg.SharedCredentialsFile {
+			credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{
+				sharedCredentialsProvider: &credentials.SharedCredentialsProvider{
+					Filename: file,
+					Profile:  cfg.Profile,
+				},
+			})
+		}
 	}
-	credProviders = append(credProviders, defaultCredProviders...)
+	credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{
+		sharedCredentialsProvider: &credentials.SharedCredentialsProvider{
+			Filename: "", Profile: "",
+		},
+	})
+	credProviders = append(credProviders, defaults.RemoteCredProvider(*config, defaults.Handlers()))
```

Comment:
If we rely on the SDK ordering as a fallback, then we don't need these or the envProvider in our chain.